File: web/src/data/mappingsReport.ts
import { config } from "../config";
import { fetchJsonWithProgress } from "./progressFetch";

export const UNRESOLVED_DIAGNOSTICS = [
  "recorded_processing_error",
  "alternative_only",
  "no_parseable_source_host",
  "no_primary_from_url_evidence",
] as const;

export const UNMAPPED_REASON_CODES = [
  "conda_cdt_repackage",
  "dependency_only_metapackage",
  "toolchain_selector",
  "environment_mutex",
  "pinning_metadata",
  "compatibility_shim",
] as const;

export type UnresolvedDiagnostic = (typeof UNRESOLVED_DIAGNOSTICS)[number];
export type UnmappedReasonCode = (typeof UNMAPPED_REASON_CODES)[number];
export type MissingPrimaryState = "explicitly_unmapped" | "unresolved";
export type UnmappedReason = {
  code: UnmappedReasonCode;
  explanation: string;
  rule_id: string;
  evidence: Record<string, string>;
};

export type MissingPrimaryPackage = {
  name: string;
  state: MissingPrimaryState;
  diagnostic_reason: UnresolvedDiagnostic | null;
  unmapped_reason: UnmappedReason | null;
  source_hosts: string[];
  version: string | null;
  source_url: string | null;
  repo: string | null;
  homepage: string | null;
  note: string | null;
  download_count: number | null;
};

export type MappingsReport = {
  schema_version: 2;
  input_schema_version: number;
  channel: string;
  counts: {
    total: number;
    primary_present: number;
    explicitly_unmapped: number;
    classified_unmapped: number;
    legacy_unmapped: number;
    unresolved: number;
    actionable_missing: number;
    primary_missing: number;
  };
  classified_by_reason: Record<UnmappedReasonCode, number>;
  unresolved_by_diagnostic: Record<UnresolvedDiagnostic, number>;
  unresolved_by_source_host: Array<{ host: string; package_count: number }>;
  missing_packages: MissingPrimaryPackage[];
};

const DIAGNOSTIC_SET = new Set<string>(UNRESOLVED_DIAGNOSTICS);
const REASON_SET = new Set<string>(UNMAPPED_REASON_CODES);
const EVIDENCE_FIELDS = new Set(["version", "build", "summary", "source_url", "repo", "homepage"]);

function record(value: unknown, label: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`${label} must be a JSON object`);
  }
  return value as Record<string, unknown>;
}

function integer(value: unknown, label: string): number {
  if (typeof value !== "number" || !Number.isInteger(value) || value < 0) {
    throw new Error(`${label} must be a non-negative integer`);
  }
  return value;
}

function nullableString(value: unknown, label: string): string | null {
  if (value === null) return null;
  if (typeof value !== "string") throw new Error(`${label} must be a string or null`);
  return value;
}

function sourceHostArray(value: unknown, label: string): string[] {
  if (!Array.isArray(value)) throw new Error(`${label} must be an array`);
  let previous: string | null = null;
  return value.map((item, index) => {
    if (
      typeof item !== "string" ||
      !item ||
      item !== item.toLowerCase() ||
      item.endsWith(".") ||
      (previous !== null && item <= previous)
    ) {
      throw new Error(`${label}[${index}] must be a normalized, unique, sorted host`);
    }
    previous = item;
    return item;
  });
}

function decodeUnmappedReason(value: unknown, label: string, state: unknown): UnmappedReason | null {
  if (value === null) return null;
  if (state !== "explicitly_unmapped") throw new Error(`${label} requires explicitly_unmapped state`);
  const reason = record(value, label);
  if (typeof reason.code !== "string" || !REASON_SET.has(reason.code)) throw new Error(`${label}.code is unsupported`);
  if (typeof reason.explanation !== "string" || !reason.explanation) throw new Error(`${label}.explanation is required`);
  if (typeof reason.rule_id !== "string" || !reason.rule_id) throw new Error(`${label}.rule_id is required`);
  const evidence = record(reason.evidence, `${label}.evidence`);
  if (Object.keys(evidence).length === 0 || Object.keys(evidence).some((key) => !EVIDENCE_FIELDS.has(key))) {
    throw new Error(`${label}.evidence is empty or unsupported`);
  }
  const decodedEvidence: Record<string, string> = {};
  for (const [key, item] of Object.entries(evidence)) {
    if (typeof item !== "string" || !item) throw new Error(`${label}.evidence.${key} is required`);
    decodedEvidence[key] = item;
  }
  return { code: reason.code as UnmappedReasonCode, explanation: reason.explanation, rule_id: reason.rule_id, evidence: decodedEvidence };
}

export function decodeMappingsReport(value: unknown): MappingsReport {
  const report = record(value, "mappings report");
  if (report.schema_version !== 2) {
    throw new Error(
      `mappings report has unsupported schema_version ${String(report.schema_version)}`,
    );
  }
  const inputSchemaVersion = integer(
    report.input_schema_version,
    "mappings report.input_schema_version",
  );
  if (typeof report.channel !== "string" || !report.channel) {
    throw new Error("mappings report.channel must be a non-empty string");
  }

  const rawCounts = record(report.counts, "mappings report.counts");
  const counts = {
    total: integer(rawCounts.total, "mappings report.counts.total"),
    primary_present: integer(
      rawCounts.primary_present,
      "mappings report.counts.primary_present",
    ),
    explicitly_unmapped: integer(
      rawCounts.explicitly_unmapped,
      "mappings report.counts.explicitly_unmapped",
    ),
    classified_unmapped: integer(rawCounts.classified_unmapped, "mappings report.counts.classified_unmapped"),
    legacy_unmapped: integer(rawCounts.legacy_unmapped, "mappings report.counts.legacy_unmapped"),
    unresolved: integer(rawCounts.unresolved, "mappings report.counts.unresolved"),
    actionable_missing: integer(rawCounts.actionable_missing, "mappings report.counts.actionable_missing"),
    primary_missing: integer(
      rawCounts.primary_missing,
      "mappings report.counts.primary_missing",
    ),
  };
  if (
    counts.primary_present + counts.explicitly_unmapped + counts.unresolved !==
      counts.total ||
    counts.explicitly_unmapped + counts.unresolved !== counts.primary_missing ||
    counts.classified_unmapped + counts.legacy_unmapped !== counts.explicitly_unmapped ||
    counts.actionable_missing !== counts.unresolved
  ) {
    throw new Error("mappings report coverage counts do not reconcile");
  }

  const rawClassifications = record(report.classified_by_reason, "mappings report.classified_by_reason");
  if (Object.keys(rawClassifications).length !== UNMAPPED_REASON_CODES.length || Object.keys(rawClassifications).some((key) => !REASON_SET.has(key))) {
    throw new Error("mappings report has unsupported classification reasons");
  }
  const classifiedByReason = Object.fromEntries(
    UNMAPPED_REASON_CODES.map((code) => [code, integer(rawClassifications[code], `mappings report.classified_by_reason.${code}`)]),
  ) as Record<UnmappedReasonCode, number>;
  if (Object.values(classifiedByReason).reduce((sum, count) => sum + count, 0) !== counts.classified_unmapped) {
    throw new Error("mappings report classification reasons do not reconcile");
  }

  const rawDiagnostics = record(
    report.unresolved_by_diagnostic,
    "mappings report.unresolved_by_diagnostic",
  );
  if (
    Object.keys(rawDiagnostics).length !== UNRESOLVED_DIAGNOSTICS.length ||
    Object.keys(rawDiagnostics).some((key) => !DIAGNOSTIC_SET.has(key))
  ) {
    throw new Error("mappings report has unsupported unresolved diagnostics");
  }
  const unresolvedByDiagnostic = Object.fromEntries(
    UNRESOLVED_DIAGNOSTICS.map((diagnostic) => [
      diagnostic,
      integer(
        rawDiagnostics[diagnostic],
        `mappings report.unresolved_by_diagnostic.${diagnostic}`,
      ),
    ]),
  ) as Record<UnresolvedDiagnostic, number>;
  if (
    Object.values(unresolvedByDiagnostic).reduce((sum, count) => sum + count, 0) !==
    counts.unresolved
  ) {
    throw new Error("mappings report unresolved diagnostics do not reconcile");
  }

  if (!Array.isArray(report.unresolved_by_source_host)) {
    throw new Error("mappings report.unresolved_by_source_host must be an array");
  }
  const unresolvedBySourceHost = report.unresolved_by_source_host.map(
    (raw, index) => {
      const label = `mappings report.unresolved_by_source_host[${index}]`;
      const item = record(raw, label);
      if (typeof item.host !== "string" || !item.host) {
        throw new Error(`${label}.host must be a non-empty string`);
      }
      return {
        host: item.host,
        package_count: integer(item.package_count, `${label}.package_count`),
      };
    },
  );
  for (let index = 0; index < unresolvedBySourceHost.length; index++) {
    const current = unresolvedBySourceHost[index];
    const previous = unresolvedBySourceHost[index - 1];
    if (
      current.host !== current.host.toLowerCase() ||
      current.host.endsWith(".") ||
      current.package_count === 0 ||
      (previous &&
        (current.package_count > previous.package_count ||
          (current.package_count === previous.package_count && current.host <= previous.host)))
    ) {
      throw new Error("mappings report unresolved source hosts are not normalized and sorted");
    }
  }

  if (!Array.isArray(report.missing_packages)) {
    throw new Error("mappings report.missing_packages must be an array");
  }
  const seenNames = new Set<string>();
  const observedDiagnostics = Object.fromEntries(
    UNRESOLVED_DIAGNOSTICS.map((diagnostic) => [diagnostic, 0]),
  ) as Record<UnresolvedDiagnostic, number>;
  const observedSourceHosts = new Map<string, number>();
  const observedClassifications = Object.fromEntries(UNMAPPED_REASON_CODES.map((code) => [code, 0])) as Record<UnmappedReasonCode, number>;
  let explicitlyUnmapped = 0;
  let unresolved = 0;
  const packages = report.missing_packages.map((raw, index): MissingPrimaryPackage => {
    const label = `mappings report.missing_packages[${index}]`;
    const pkg = record(raw, label);
    if (typeof pkg.name !== "string" || !pkg.name || seenNames.has(pkg.name)) {
      throw new Error(`${label}.name must be a unique non-empty string`);
    }
    seenNames.add(pkg.name);
    if (pkg.state !== "explicitly_unmapped" && pkg.state !== "unresolved") {
      throw new Error(`${label}.state is unsupported`);
    }
    let diagnostic: UnresolvedDiagnostic | null = null;
    if (pkg.state === "unresolved") {
      if (typeof pkg.diagnostic_reason !== "string" || !DIAGNOSTIC_SET.has(pkg.diagnostic_reason)) {
        throw new Error(`${label}.diagnostic_reason is required for unresolved packages`);
      }
      diagnostic = pkg.diagnostic_reason as UnresolvedDiagnostic;
      observedDiagnostics[diagnostic]++;
      unresolved++;
    } else {
      if (pkg.diagnostic_reason !== null) {
        throw new Error(`${label}.diagnostic_reason must be null when explicitly unmapped`);
      }
      explicitlyUnmapped++;
    }
    const unmappedReason = decodeUnmappedReason(pkg.unmapped_reason, `${label}.unmapped_reason`, pkg.state);
    if (unmappedReason) observedClassifications[unmappedReason.code]++;
    const sourceHosts = sourceHostArray(pkg.source_hosts, `${label}.source_hosts`);
    if (pkg.state === "unresolved") {
      for (const host of sourceHosts) {
        observedSourceHosts.set(host, (observedSourceHosts.get(host) ?? 0) + 1);
      }
    }
    const downloads = pkg.download_count;
    if (downloads !== null && (typeof downloads !== "number" || !Number.isInteger(downloads) || downloads < 0)) {
      throw new Error(`${label}.download_count must be a non-negative integer or null`);
    }
    return {
      name: pkg.name,
      state: pkg.state,
      diagnostic_reason: diagnostic,
      unmapped_reason: unmappedReason,
      source_hosts: sourceHosts,
      version: nullableString(pkg.version, `${label}.version`),
      source_url: nullableString(pkg.source_url, `${label}.source_url`),
      repo: nullableString(pkg.repo, `${label}.repo`),
      homepage: nullableString(pkg.homepage, `${label}.homepage`),
      note: nullableString(pkg.note, `${label}.note`),
      download_count: downloads as number | null,
    };
  });
  if (
    packages.length !== counts.primary_missing ||
    explicitlyUnmapped !== counts.explicitly_unmapped ||
    unresolved !== counts.unresolved ||
    UNMAPPED_REASON_CODES.some((code) => observedClassifications[code] !== classifiedByReason[code]) ||
    UNRESOLVED_DIAGNOSTICS.some(
      (diagnostic) => observedDiagnostics[diagnostic] !== unresolvedByDiagnostic[diagnostic],
    ) ||
    unresolvedBySourceHost.length !== observedSourceHosts.size ||
    unresolvedBySourceHost.some(
      ({ host, package_count }) => observedSourceHosts.get(host) !== package_count,
    )
  ) {
    throw new Error("mappings report package rows do not reconcile with aggregate counts");
  }

  return {
    schema_version: 2,
    input_schema_version: inputSchemaVersion,
    channel: report.channel,
    counts,
    classified_by_reason: classifiedByReason,
    unresolved_by_diagnostic: unresolvedByDiagnostic,
    unresolved_by_source_host: unresolvedBySourceHost,
    missing_packages: packages,
  };
}

export async function loadMappingsReport(
  path = config.mappingsReportUrl,
): Promise<MappingsReport> {
  return decodeMappingsReport(
    await fetchJsonWithProgress<unknown>(path, { cache: "no-cache" }),
  );
}

File: scripts/mappings_report.py
"""Report primary-PURL coverage from a freshly merged full mapping bundle.

This is a read-only report, not a mapper or a source-layer merger. Rebuild older
published bundles with scripts.merge_mappings first: the current identities
contract is required. Alternative PURLs and CPEs do not count as primary PURLs.
"""

from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any
from urllib.parse import urlsplit

from scripts import merge_mappings, validate

REPORT_SCHEMA_VERSION = 1
URL_EVIDENCE_FIELDS = ("source_url", "repo", "homepage")
EVIDENCE_FIELDS = ("version", *URL_EVIDENCE_FIELDS, "note")
UNRESOLVED_DIAGNOSTICS = (
    "recorded_processing_error",
    "alternative_only",
    "no_parseable_source_host",
    "no_primary_from_url_evidence",
)


def _source_hosts(entry: dict[str, Any]) -> tuple[str, ...]:
    """Return normalized, unique hostnames from persisted URL evidence."""
    hosts: set[str] = set()
    for field in URL_EVIDENCE_FIELDS:
        value = entry.get(field)
        if not isinstance(value, str) or not value.strip():
            continue
        try:
            hostname = urlsplit(value).hostname
        except ValueError:
            continue
        if hostname:
            normalized = hostname.lower().rstrip(".")
            if normalized:
                hosts.add(normalized)
    return tuple(sorted(hosts))


def _has_alternative_purl(entry: dict[str, Any]) -> bool:
    """Use the canonical identity contract rather than legacy alternatives."""
    return any(
        identity["kind"] == "purl" and identity["role"] == "alternative"
        for identity in entry["identities"]
    )


def _diagnose_unresolved(entry: dict[str, Any]) -> str:
    """Classify persisted evidence without claiming an authoritative root cause."""
    note = (entry.get("note") or "").strip()
    if note.lower().startswith("fetch error:"):
        return "recorded_processing_error"
    if _has_alternative_purl(entry):
        return "alternative_only"
    if not _source_hosts(entry):
        return "no_parseable_source_host"
    return "no_primary_from_url_evidence"


def build_report(payload: dict[str, Any]) -> dict[str, Any]:
    """Summarize effective identities without inferring why a PURL is absent.

    Explicit unmapped decisions are distinct from unresolved packages. Neither
    missing URLs nor automatic diagnostic notes establish an intentional
    no-PURL decision. Notes can survive reviewed overrides and are evidence,
    not authoritative mapping state.
    """
    if not isinstance(payload, dict):
        raise ValueError("bundle must be an object")
    if (
        type(payload.get("schema_version")) is not int
        or payload["schema_version"] != validate.CURRENT_BUNDLE_SCHEMA
    ):
        raise ValueError(
            f"report requires full bundle schema {validate.CURRENT_BUNDLE_SCHEMA}; "
            "rebuild with scripts.merge_mappings"
        )
    packages = payload.get("packages")
    if not isinstance(packages, dict):
        raise ValueError("bundle.packages must be an object")
    if type(payload.get("package_count")) is not int or payload["package_count"] != len(
        packages
    ):
        raise ValueError("bundle.package_count must equal the number of packages")
    channel = payload.get("channel")
    if not isinstance(channel, str) or not channel:
        raise ValueError("bundle.channel must be a non-empty string")
    if any(not isinstance(name, str) or not name for name in packages):
        raise ValueError("package names must be non-empty strings")

    counts = {"primary_present": 0, "explicitly_unmapped": 0, "unresolved": 0}
    unresolved_by_diagnostic = {diagnostic: 0 for diagnostic in UNRESOLVED_DIAGNOSTICS}
    missing: list[dict[str, Any]] = []
    for name, entry in sorted(packages.items()):
        label = f"packages.{name}"
        if not isinstance(entry, dict):
            raise ValueError(f"{label} must be an object")
        if "unmapped" in entry and not isinstance(entry["unmapped"], bool):
            raise ValueError(f"{label}.unmapped must be a boolean")
        if entry.get("status") == "unmapped" and entry.get("unmapped") is not True:
            raise ValueError(f"{label}.status unmapped requires unmapped: true")
        if entry.get("purl") is not None and not isinstance(entry["purl"], str):
            raise ValueError(f"{label}.purl must be a string or null")
        for field in EVIDENCE_FIELDS:
            if entry.get(field) is not None and not isinstance(entry[field], str):
                raise ValueError(f"{label}.{field} must be a string or null")
        downloads = entry.get("download_count")
        if downloads is not None and (type(downloads) is not int or downloads < 0):
            raise ValueError(
                f"{label}.download_count must be a non-negative integer or null"
            )

        # Reuse the published contract, including legacy/identity consistency
        # and reviewed-unmapped invariants, rather than accepting corrupt data
        # as apparently successful empty coverage.
        errors: list[str] = []
        validate._validate_identity_contract(entry, label, errors)
        if errors:
            raise ValueError("\n".join(errors))
        has_primary = any(
            identity["kind"] == "purl" and identity["role"] == "primary"
            for identity in entry["identities"]
        )
        state = (
            "primary_present"
            if has_primary
            else "explicitly_unmapped"
            if entry.get("unmapped") is True
            else "unresolved"
        )
        counts[state] += 1
        if not has_primary:
            diagnostic_reason = None
            if state == "unresolved":
                diagnostic_reason = _diagnose_unresolved(entry)
                unresolved_by_diagnostic[diagnostic_reason] += 1
            missing.append(
                {
                    "name": name,
                    "state": state,
                    "diagnostic_reason": diagnostic_reason,
                    **{field: entry.get(field) for field in EVIDENCE_FIELDS},
                    "download_count": downloads,
                }
            )

    if sum(counts.values()) != len(packages):
        raise RuntimeError("primary-PURL coverage counts do not reconcile")
    if sum(unresolved_by_diagnostic.values()) != counts["unresolved"]:
        raise RuntimeError("unresolved diagnostic counts do not reconcile")

    return {
        "schema_version": REPORT_SCHEMA_VERSION,
        "input_schema_version": payload["schema_version"],
        "channel": channel,
        "counts": {
            "total": len(packages),
            **counts,
            "primary_missing": counts["explicitly_unmapped"] + counts["unresolved"],
        },
        "unresolved_by_diagnostic": unresolved_by_diagnostic,
        "missing_packages": missing,
    }


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--input",
        type=Path,
        required=True,
        help="Freshly merged full mappings.json bundle",
    )
    parser.add_argument(
        "--output",
        type=Path,
        help="Write the report to this path instead of standard output",
    )
    args = parser.parse_args()
    try:
        # The shared loader rejects duplicate keys and non-finite JSON numbers.
        report = build_report(merge_mappings._load_json(args.input))
        rendered = json.dumps(report, indent=2, sort_keys=True, allow_nan=False) + "\n"
        if args.output:
            args.output.parent.mkdir(parents=True, exist_ok=True)
            args.output.write_text(rendered)
        else:
            print(rendered, end="")
    except (OSError, ValueError) as exc:
        parser.exit(1, f"mappings report: {exc}\n")


if __name__ == "__main__":
    main()
